examples/flutter_gallery/lib/demo/pan_and_zoom_demo_transform_interaction.dart

```diff
@@ -0,0 +1,460 @@
+    if (widget.disableScale) {
+      return matrix;
+    }
+
+    // Don't allow a scale that moves the viewport outside of _visibleRect.
+    final Offset tl = fromViewport(const Offset(0, 0), _transform);
+    final Offset tr = fromViewport(Offset(widget.size.width, 0), _transform);
+    final Offset bl = fromViewport(Offset(0, widget.size.height), _transform);
+    final Offset br = fromViewport(
+      Offset(widget.size.width, widget.size.height),
+      _transform,
+    );
+    if (!_visibleRect.contains(tl)
+      || !_visibleRect.contains(tr)
+      || !_visibleRect.contains(bl)
+      || !_visibleRect.contains(br)) {
+      return matrix;
+    }
+
+    // Don't allow a scale that results in an overall scale beyond min/max scale
+    final double currentScale = _transform.getMaxScaleOnAxis();
+    final double totalScale = currentScale * scale;
+    final double clampedTotalScale = totalScale.clamp(
+      widget.minScale,
+      widget.maxScale,
+    );
+    final double clampedScale = clampedTotalScale / currentScale;
+    return matrix..scale(clampedScale);
+  }
+
+  // Return a new matrix representing the given matrix after applying the given
+  // rotation transform.
+  // Rotating the scene cannot cause the viewport to view beyond _visibleRect.
+  Matrix4 matrixRotate(Matrix4 matrix, double rotation, Offset focalPoint) {
+    if (widget.disableRotation) {
+      return matrix;
+    }
+    final Offset focalPointScene = fromViewport(focalPoint, matrix);
+    return matrix
+      ..translate(focalPointScene.dx, focalPointScene.dy)
+      ..rotateZ(-rotation)
+      ..translate(-focalPointScene.dx, -focalPointScene.dy);
+  }
+
+  // Handle panning and pinch zooming events
+  void _onScaleStart(ScaleStartDetails details) {
+    widget.onScaleStart?.call();
+
+    // TODO(justinmc): Do things like this need to happen inside of setState? I
+    // don't need build to be called because of any of this.
+    if (_controller.isAnimating) {
+      _controller.stop();
+      _controller.reset();
+      _animation?.removeListener(_onAnimate);
+      _animation = null;
+    }
+    if (_controllerReset.isAnimating) {
+      _controllerReset.stop();
+      _animationReset?.removeListener(_onAnimateReset);
+      _animationReset = null;
+      _controllerReset.reset();
+      widget.onResetEnd();
+    }
+
+    gestureType = null;
+    setState(() {
+      _scaleStart = _transform.getMaxScaleOnAxis();
+      _translateFromScene = fromViewport(details.focalPoint, _transform);
+      _rotationStart = _currentRotation;
+    });
+  }
+  void _onScaleUpdate(ScaleUpdateDetails details) {
+    widget.onScaleUpdate?.call(fromViewport(details.focalPoint, _transform));
+    double scale = _transform.getMaxScaleOnAxis();
+    final Offset focalPointScene = fromViewport(
+      details.focalPoint,
+      _transform,
+    );
+    if (gestureType == null) {
+      if ((details.scale - 1).abs() > details.rotation.abs()) {
```

Comment:
I'll write a comment for this.  I disallow multiple types of transforms in the same gesture, so this line decides whether the gesture is a scale or a rotation.  It does this by comparing whether the gesture has more scale or more rotation in it.  The `-1` is because scale starts at 1 but rotation starts at 0.